### PR TITLE
Handle native backend tensors in data adapters

### PR DESCRIPTION
**Pull request by samudraneel05**
## Problem

Pluggable backend tensors (e.g. MLX arrays) are rejected by the array data adapter because `can_slice_array()` and `convert_to_sliceable()` only check for built-in backend types (numpy, TF, JAX, torch, pandas, scipy) and `__array__`. MLX intentionally removed `__array__` in favor of the Python Buffer Protocol ([ml-explore/mlx#323](https://github.com/ml-explore/mlx/pull/323)), so MLX arrays fall through every existing check and `model.fit/evaluate/predict` crash.

## Fix

Add `ops.is_tensor(x)` as a fallback in both `can_slice_array()` and `convert_to_sliceable()`. This is the approach suggested in [keras-mlx#40](https://github.com/keras-team/keras-mlx/pull/40): the data adapters should handle native arrays where `is_tensor(x)` is true.

For the sliceable representation, use `np.asarray(x)` to get a zero-copy numpy view via the buffer protocol (MLX on Apple Silicon unified memory shares memory with numpy without a copy). Fall back to `ops.convert_to_numpy(x)` for dtypes that `np.asarray` can't handle directly — the only known case is MLX bfloat16, which fails the buffer protocol because numpy doesn't natively understand bfloat16. PyTorch bfloat16 has the same issue with `np.asarray`.

## Approaches considered

**1. On-device MLX slicing (native `Sliceable`)**
Keep arrays as MLX, use `mx.slice`/`mx.take` or direct indexing for batches. Rejected because:
- MLX slicing creates a copy, not a view (unlike numpy)
- MLX doesn't accept numpy index arrays (`mx_array[np_array]` fails), and Keras's shuffle path produces numpy permutations
- MLX's gather is lazy; in Keras's data adapter (which yields concrete batches) each gather becomes a separate graph evaluation. Benchmark: 54ms vs 34ms for the numpy path
- MLX's idiomatic pattern (see [MLP example](https://ml-explore.github.io/mlx/build/html/examples/mlp.html)) keeps gathers lazy and fuses them with the training step — Keras's data adapter architecture can't do this

**2. Always `ops.convert_to_numpy` (Keras Hub pattern)**
keras-hub's `tensor_utils.convert_to_numpy` uses `ops.is_tensor(x)` → `ops.convert_to_numpy(x)` preemptively. Simple, but always copies the full dataset. This is the overhead the maintainer wanted to avoid.

**3. Zero-copy `np.asarray` with fallback (this PR)**
Try `np.asarray` first (zero-copy view via buffer protocol), fall back to `ops.convert_to_numpy` only when it fails. Eliminates the copy for all dtypes except bfloat16.

Benchmark (50k×256 float32, 5 `fit` calls):
```
convert_to_numpy (old): 29.8ms
np.asarray (new):        0.0ms
```

## bfloat16 handling

`np.asarray(mx_bfloat16)` raises `RuntimeError` (buffer format mismatch). The fallback `ops.convert_to_numpy` preserves bfloat16 correctly, and the existing floatx normalization step (lines 421-424) casts it to floatx. Verified end-to-end with `model.fit` on bfloat16 inputs.

Tested all 13 MLX dtypes — only bfloat16 triggers the fallback.

## Test plan

- [x] `test_fit_with_data_adapter_native_array` passes (the test un-excluded in keras-mlx#40)
- [x] All 71 existing `array_data_adapter_test.py` tests pass (no regressions)
- [x] All 133 `trainer_test.py` tests pass with mlx backend
- [x] All 10 `epoch_iterator_test.py` tests pass
- [x] `model.fit/evaluate/predict` with MLX array inputs (float32, bfloat16, nested dict, shuffle, validation_split)

### CODE REVIEW

**Comment by samudraneel05** [comment]
Closing, will reopen later.
